examples/dataflow-python-examples/streaming-examples/nlp-sentiment-analysis-pipeline/src/beam/pipeline.py

```diff
@@ -0,0 +1,413 @@
+        return dict(merged_accumulator)
+
+    def extract_output(self, accumulator):
+        return accumulator
+
+
+class AddWindowingInformation(beam.DoFn):
+    """Adds start time of window to each element.
+
+    Only used when --nlp_batching is set. By adding the window's
+    start time to the key of each element, the pipeline
+    can group each element per window per key into a document.
+    Additionally, after 500 tweets per date, the document version
+    will be incremented to avoid overloading the NLP API.
+
+    TWEET_INDEX and DOC_INDEX are stateful dictionaries that
+    track how many tweets are in each document and what version
+    the respective document is on. After 500 tweets per day,
+    a new document should be created and the version should
+    consequently be incremented.
+
+    For example, a modified key might be: "2020-05-20+XXXXXXXXXX+3"
+    """
+    TWEET_INDEX = beam.transforms.userstate.CombiningValueStateSpec('tweet_index', DictCombineFn())
+    DOC_INDEX = beam.transforms.userstate.CombiningValueStateSpec('doc_index', DictCombineFn())
+
+    def process(self, tweet_tuple, tweet_index=beam.DoFn.StateParam(TWEET_INDEX), doc_index=beam.DoFn.StateParam(DOC_INDEX), window=beam.DoFn.WindowParam):
+        time_value, tweet_text = tweet_tuple
+
+        tweet_index.add([time_value, 1])
+
+        if tweet_index.read()[time_value] > 500:
+            tweet_index.clear()
+            doc_index.add([time_value, 1])
+
+        window_start = int(window.start.to_utc_datetime().timestamp())
+        doc_index = doc_index.read().get(time_value, 0)
+
+        modified_key = "{time_value}+{window_start}+{doc_index}".format(time_value=time_value, window_start=window_start, doc_index=doc_index)
+        yield (modified_key, tweet_text)
+
+
+class NlpClient_Batched(beam.DoFn):
+    """Gets sentiment score for each tweet via batching.
+
+    This class batches each tweet into a document. There is
+    one document per window per key. In other words, if there
+    exists 3 dates within a window, there are 3 documents made.
+    The document approach is meant to drastically reduce the
+    number of API calls to the Natural Language API.
+
+    Attributes:
+      language: Language of tweets.
+      client: NLP API client instance.
+    """
+    def __init__(self, client=None):
+        self.language = "en"
+        self.client = client
+
+    def setup(self): # NOTE: setup() is only supported with Dataflow Runner or Direct Runner (Apache Beam 2.22.0)
+        if self.client is None:
+            self.client = language.LanguageServiceClient() 
+
+    def process(self, modified_tweet_tuple):
+        if self.client is None:
+            self.client = language.LanguageServiceClient()
+
+        (modified_key, text_iterable) = modified_tweet_tuple
+
+        time_value = modified_key[:modified_key.index("+")]
+
+        file_name = './' + modified_key + ".txt"
+        writer = filesystems.FileSystems.create(file_name)
```

Comment:
There is no reason to use file system here as the scope of this "file" is just a single process element call and you ultimately just read the whole thing back into a string anyway.

It would be much faster (and more concise) to provide an in memory file-like object. Take a look at [StringIO](https://docs.python.org/3/library/io.html#io.StringIO) and use a [context manager](https://book.pythontips.com/en/latest/context_managers.html) for opening / closing.
Looks something like this:
```python3
import io
with open(io.StringIO) as tweet_doc:
    ...
    for tweet in test_iterable:
        tweet_doc.write(...)
    self.analyze_document(tweet_doc) # This should be refactored to take file-like object rather than file name.
```